go/chat/signencrypt/seeker.go

```diff
@@ -0,0 +1,177 @@
+		r.chunks.Add(c.index, pt[c.ptStart-start:c.ptEnd-start])
+	}
+}
+
+func (r *decodingReadSeeker) fetchChunks(chunks []chunkSpec) (res []byte, err error) {
+	// we want to fetch enough data for all the chunks in one hit on the source ReadSeeker
+	begin := chunks[0].cipherStart
+	end := chunks[len(chunks)-1].cipherEnd
+	num := end - begin
+
+	if _, err := r.source.Seek(begin, io.SeekStart); err != nil {
+		return res, err
+	}
+	var bufOffset int64
+	res = make([]byte, num)
+	for {
+		n, err := r.source.Read(res[bufOffset:])
+		if err != nil {
+			return res, err
+		}
+		bufOffset += int64(n)
+		if bufOffset >= num {
+			break
+		}
+	}
+	return res, nil
+}
+
+func (r *decodingReadSeeker) clamp(offset int64) int64 {
+	if offset >= r.size {
+		offset = r.size
+	}
+	return offset
+}
+
+func (r *decodingReadSeeker) extractPlaintext(plainText []byte, num int64, chunks []chunkSpec) []byte {
+	datBegin := chunks[0].ptStart
+	ptBegin := r.offset
+	ptEnd := r.clamp(r.offset + num)
+	r.Debug(r.ctx, "extractPlaintext: datBegin: %v ptBegin: %v ptEnd: %v", datBegin, ptBegin, ptEnd)
+	return plainText[ptBegin-datBegin : ptEnd-datBegin]
+}
+
+func (r *decodingReadSeeker) Read(res []byte) (n int, err error) {
+	defer r.Trace(r.ctx, func() error { return err }, "Read(%v,%v)", r.offset, len(res))()
+	if r.offset >= r.size {
+		return 0, io.EOF
+	}
+	num := int64(len(res))
+	chunkEnd := r.clamp(r.offset + num)
+	r.Debug(r.ctx, "Read: chunkEnd: %v", chunkEnd)
+	chunks := getChunksInRange(r.offset, chunkEnd, r.size)
+	var chunkPlaintext []byte
+
+	// Check for a full hit on all the chunks first
+	var ok bool
+	if chunkPlaintext, ok = r.getChunksFromCache(chunks); !ok {
+		// if we miss, then we need to fetch the data from our underlying source. Given that this
+		// source is usually on the network, then fetch at least 4 chunks so we aren't making
+		// too many requests.
+		minChunkEnd := r.clamp(r.offset + 4*DefaultPlaintextChunkLength)
+		if minChunkEnd > chunkEnd {
+			chunkEnd = minChunkEnd
+		}
+		prefetchChunks := getChunksInRange(r.offset, chunkEnd, r.size)
+		cipherText, err := r.fetchChunks(prefetchChunks)
+		if err != nil {
+			return n, err
+		}
+		for _, c := range prefetchChunks {
+			r.Debug(r.ctx, "Read: chunk: index: %v ptstart: %v ptend: %v cstart: %v cend: %v", c.index,
+				c.ptStart, c.ptEnd, c.cipherStart, c.cipherEnd)
+		}
+		// Decrypt all the chunks and write out to the cache
+		decoder := NewDecoder(r.encKey, r.verifyKey, r.sigPrefix, r.nonce)
+		decoder.setChunkNum(uint64(prefetchChunks[0].index))
+		if chunkPlaintext, err = decoder.Write(cipherText); err != nil {
+			return n, err
+		}
+		// We might have some straggling data, so just hit Finish here to potentially pick it up. If it
+		// returns an error, then we just ignore it.
+		if finishPlaintext, err := decoder.Finish(); err == nil {
+			chunkPlaintext = append(chunkPlaintext, finishPlaintext...)
+		}
+		r.writeChunksToCache(chunkPlaintext, prefetchChunks)
+	}
```

Comment:
I wonder if it would also help to fetch more chunks on a cache hit.  i.e. if you prefetched chunks 0, 1, 2, 3 and just hit on chunk 2, then fetch 4, 5 now while reading 2.

Reply:
The problem there is it makes every cache hit slow. I was thinking about how I could run this in the background, but it was going to get complicated. So the flow is basically this (on Darwin at least):

1.) Player requests the file to get MIME type, we download 4MB. 
2.) Player requests bytes 0-1 for reasons I don't understand, we hit the 4MB chunk from before.
3.) Player requests most of the rest of the file (but probably not all of it). We start churning through the stream as fast as we can in 4MB increments. The cache hits are so fast it doesn't really matter if we start the download on one of them, since we will eventually hit the end of the cache and start up the download while playback is happening.

Reply:
I found that the ideal number is dependent on the size of the video as well. I tried streaming a 2GB file last night and it didn't really work. For a bigger video I think it makes sense to use a bigger value for this, I will try that this morning.

Reply:
I was thinking that the fetch (when there's a hit) would be in the background.  Just a thought, but no need whatsoever to pursue that path.

Reply:
Yeah I was thinking about how to do it in the background, but the whole story seemed to get more complicated that way. Maybe something we can look at in the future though.